File: src/test/java/org/albaross/agents4j/extraction/utils/InputGenerator.java
package org.albaross.agents4j.extraction.utils;

import org.albaross.agents4j.extraction.KnowledgeBase;
import org.albaross.agents4j.extraction.data.Pair;
import org.albaross.agents4j.extraction.extractors.AdaptedExtractor;

import java.util.*;

public class InputGenerator {

    private static final Random RND = new Random();

    public static final <A> Collection<Pair<A>> generate(int count, List<String> dimensions, List<A> actions) {

        if (count <= 0 || dimensions.isEmpty() || actions.isEmpty())
            return Collections.emptySet();

        int range = (int) Math.ceil(Math.pow(50 * count, (double) 1 / dimensions.size()));
        Set<Pair<A>> pairs = new HashSet<>(count);

        while (pairs.size() < count) {
            Pair.PairBuilder<A> builder = Pair.<A>builder();
            for (String s : dimensions)
                builder.symbol(s + "_" + RND.nextInt(range));

            builder.action(actions.get(RND.nextInt(actions.size())));
            pairs.add(builder.build());
        }

        return pairs;
    }

    public static void main(String... args) {
        Collection<Pair<String>> pairs = InputGenerator.generate(1000, Arrays.asList("u", "t", "x", "y", "z"), Arrays.asList("north", "south", "east", "west"));
        System.out.println("generated pairs");
        AdaptedExtractor<String> ext = new AdaptedExtractor<>();
        long start, end;
        KnowledgeBase<String> kb;
        start = System.currentTimeMillis();
        kb = ext.apply(pairs);
        end = System.currentTimeMillis();
        System.out.println(kb);
        System.out.println((end - start) + "ms");
    }

}

File: src/test/java/InputGenerator.kt
import org.albaross.agents4j.extraction.KnowledgeBase
import org.albaross.agents4j.extraction.bases.MultiAbstractionLevelKB
import org.albaross.agents4j.extraction.data.Pair
import org.albaross.agents4j.extraction.extractors.MemExtractor
import org.albaross.agents4j.extraction.utils.format
import java.util.*
import java.util.Collections.emptySet
import java.util.function.Supplier

private val RND = Random()

fun <A> generate(count: Int, dimensions: List<String>, actions: List<A>): Collection<Pair<A>> {

    if (count <= 0 || dimensions.isEmpty() || actions.isEmpty())
        return emptySet()

    val range = Math.ceil(Math.pow((50 * count).toDouble(), 1.toDouble() / dimensions.size)).toInt()
    val pairs = TreeSet<Pair<A>>(kotlin.Comparator { o1, o2 -> o1.state.toString().compareTo(o2.state.toString()) })

    while (pairs.size < count) {
        val state = TreeSet<String>()
        for (s in dimensions) {
            state.add("${s}_${RND.nextInt(range)}")
        }

        val action = actions[RND.nextInt(actions.size)]
        pairs.add(Pair(state, action))
    }

    return pairs
}

fun main(args: Array<String>) {
    val pairs = generate(5000, listOf("s", "t", "u", "v"), listOf("north", "south", "east", "west"))
    println("generated pairs")
    val ext = MemExtractor<String>(Supplier { MultiAbstractionLevelKB<String>() })
    val kb: KnowledgeBase<String>
    val start = System.currentTimeMillis()
    kb = ext.apply(pairs)
    val end = System.currentTimeMillis()
    println(kb.format())
    println((end - start).toString() + "ms")
}